Mathlib/Data/Nat/ModEq.lean

```diff
@@ -530,4 +543,129 @@ theorem odd_mod_four_iff {n : ℕ} : n % 2 = 1 ↔ n % 4 = 1 ∨ n % 4 = 3 :=
+      (by simp; intro j hj; exact (List.pairwise_cons.mp co).1 j hj)
+    exact chineseRemainder_modEq_unique this
+      (hz i (by simp)) (ih co.of_cons (fun j hj => hz j (by simp [hj])))
+
+theorem chineseRemainderOfList_perm {l l' : List ι} (hl : l.Perm l')
+    (hs : ∀ i ∈ l, s i ≠ 0) (co : l.Pairwise (Coprime on s)) :
+    (chineseRemainderOfList a s l co : ℕ) =
+    chineseRemainderOfList a s l' (co.perm hl coprime_comm.mpr) := by
+  let z := chineseRemainderOfList a s l' (co.perm hl coprime_comm.mpr)
+  have hlp : (l.map s).prod = (l'.map s).prod := List.Perm.prod_eq (List.Perm.map s hl)
+  exact (chineseRemainderOfList_modEq_unique a s l co (z := z)
+    (fun i hi => z.prop i (hl.symm.mem_iff.mpr hi))).symm.eq_of_lt_of_lt
+      (chineseRemainderOfList_lt_prod _ _ _ _ hs)
+      (by rw [hlp]
+          exact chineseRemainderOfList_lt_prod _ _ _ _
+            (by simpa [List.Perm.mem_iff hl.symm] using hs))
+
+/-- The natural number less than `(m.map s).prod` congruent to
+`a i` mod `s i` for all  `i ∈ m`. -/
+def chineseRemainderOfMultiset {m : Multiset ι} :
+    m.Nodup → (∀ i ∈ m, s i ≠ 0) → Set.Pairwise {x | x ∈ m} (Coprime on s) →
+    { k // ∀ i ∈ m, k ≡ a i [MOD s i] } :=
+  Quotient.recOn m
+    (fun l nod _ co =>
+      chineseRemainderOfList a s l (List.Nodup.pairwise_of_forall_ne nod co))
+    (fun l l' (pp : l.Perm l') ↦
+      funext fun nod' : l'.Nodup =>
+      have nod : l.Nodup := pp.symm.nodup_iff.mp nod'
+      funext fun hs' : ∀ i ∈ l', s i ≠ 0 =>
+      have hs : ∀ i ∈ l, s i ≠ 0  := by simpa [List.Perm.mem_iff pp] using hs'
+      funext fun co' : Set.Pairwise {x | x ∈ l'} (Coprime on s) =>
+      have co : Set.Pairwise {x | x ∈ l} (Coprime on s) := by simpa [List.Perm.mem_iff pp] using co'
+      have lco : l.Pairwise (Coprime on s) := List.Nodup.pairwise_of_forall_ne nod co
+      have : ∀ {m' e nod'' hs'' co''}, @Eq.ndrec (Multiset ι) l
+        (fun m ↦ m.Nodup → (∀ i ∈ m, s i ≠ 0) →
+          Set.Pairwise {x | x ∈ m} (Coprime on s) → { k // ∀ i ∈ m, k ≡ a i [MOD s i] })
+        (fun nod _ co ↦ chineseRemainderOfList a s l (List.Nodup.pairwise_of_forall_ne nod co))
+          m' e nod'' hs'' co'' =
+        (chineseRemainderOfList a s l lco : ℕ) := by
+          rintro _ rfl _ _ _; rfl
+      by ext; exact this.trans <| chineseRemainderOfList_perm a s pp hs lco)
+
+theorem chineseRemainderOfMultiset_lt_prod {m : Multiset ι}
+    (nod : m.Nodup) (hs : ∀ i ∈ m, s i ≠ 0) (pp : Set.Pairwise {x | x ∈ m} (Coprime on s)) :
+    chineseRemainderOfMultiset a s nod hs pp < (m.map s).prod := by
+  induction' m using Quot.ind with l
+  unfold chineseRemainderOfMultiset; simp
+  exact chineseRemainderOfList_lt_prod a s l
+    (List.Nodup.pairwise_of_forall_ne nod pp) (by simpa using hs)
+
+/-- The natural number less than `∏ i in t, s i` congruent to
+`a i` mod `s i` for all  `i ∈ t`. -/
+def chineseRemainderOfFinset (t : Finset ι)
+    (hs : ∀ i ∈ t, s i ≠ 0) (pp : Set.Pairwise t (Coprime on s)) :
```

Comment:
Again, would it be better to separate out the data that's the output from the proof it works?